enhancements/installer/kubernete-slb-owned-api-loadbalancers.md

```diff
@@ -0,0 +1,508 @@
+---
+title: kubernetes-slb-owned-api-loadbalancers
+authors:
+  - "@abhinavdahiya"
+reviewers:
+  - "@sttts"
+  - "@smarterclayton"
+approvers:
+  - "@sttts"
+  - "@smarterclayton"
+creation-date: 2020-05-06
+last-updated: 2020-08-26
+status: implementable
+see-also:
+  -   
+replaces:
+  - 
+superseded-by:
+  - 
+---
+
+# Kubernetes Service Owned API Loadbalancers
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+1. Service type Loadbalancer's ExternalTrafficPolicy should be Cluster or Local ?
+2. How to reserve private IPs on AWS ?
+
```

Comment:
There is one crucial detail because we want to self-host:

**Will the cloud provider code mirror the pod readiness probe into the load balancer object? I.e. will the gcp/azure/AWS/whatever LB objects have real health checks?**

I ask this because there are also probes of Exec type which will certainly not be mirrored. So I fear this logic is on the cluster-side and endpoints are synched with the cloud provider from some in-cluster LB backend service.

If this probing is in cluster, I believe we run into another hard-to-understand cyclic dependency because this in-cluster LB backend cannot speak to the API because the LBs endpoints are not up-to-date.

@frobware can you find out? This has blocker potential for the whole approach.

Reply:
Note: this is a little similar to our static pods today. We can't rely on the control plane to be up to be able to run the control plane. Static pods solve that by writing disk files making the static pods independent of themselves. Same is necessary for LBs: they must be "self-standing" as soon as they are configured by the in-cluster LB backend once.